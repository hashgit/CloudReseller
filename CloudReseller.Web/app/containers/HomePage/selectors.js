/**
 * Homepage selectors
 */

import { createSelector } from 'reselect';

const selectHome = (state) => state.get('home');

const makeSelectProducts = () => createSelector(
  selectHome,
  (homeState) => homeState.get('products').toJS()
);

export {
  selectHome,
  makeSelectProducts,
};
