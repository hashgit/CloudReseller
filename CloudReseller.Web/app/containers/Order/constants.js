export const ADD_PRODUCT = 'boilerplate/Home/ADD_PRODUCT';
export const PLACE_ORDER = 'boilerplate/Home/PLACE_ORDER';
export const PLACED = 'boilerplate/Home/PLACED';
