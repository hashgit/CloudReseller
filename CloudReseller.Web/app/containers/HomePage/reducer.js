/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */
import { fromJS } from 'immutable';

import { CHANGE_USERNAME, PRODUCTS_LOADED, LOAD_PRODUCTS, DISPLAY_CURRENCY } from './constants';

// The initial state of the App
const initialState = fromJS({
  username: '',
  currency: '',
  displayCurrency: '',
  products: [],
});

function homeReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_PRODUCTS:
      return state.set('currency', action.currency);
    case DISPLAY_CURRENCY:
      return state.set('displayCurrency', action.currency);
    case PRODUCTS_LOADED:
      return state.set('products', fromJS(action.products));
    case CHANGE_USERNAME:
      // Delete prefixed '@' from the github username
      return state.set('username', action.name.replace(/@/gi, ''));
    default:
      return state;
  }
}

export default homeReducer;
