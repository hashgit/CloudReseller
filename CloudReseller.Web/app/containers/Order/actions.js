import { ADD_PRODUCT, PLACE_ORDER, PLACED } from './constants';

export function addProduct(product) {
  return {
    type: ADD_PRODUCT,
    product
  };
}

export function placeOrder(data) {
  return {
    type: PLACE_ORDER,
    data
  };
}

export function placed() {
  return {
    type: PLACED,
  };
}
