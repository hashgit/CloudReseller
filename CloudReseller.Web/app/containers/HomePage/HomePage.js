/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import ReposList from 'components/ReposList';
import Order from 'containers/Order';
import './style.scss';

export default class HomePage extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  /**
   * when initial state username is not null, submit the form to load repos
   */
  componentDidMount() {
    this.props.loadProducts();
  }

  render() {
    const { loading, error, products } = this.props;
    const reposListProps = {
      loading,
      error,
      products,
    };

    return (
      <article>
        <Helmet>
          <title>Home Page</title>
          <meta name="description" content="A React.js Boilerplate application homepage" />
        </Helmet>
        <div className="home-page">
          <section className="centered">
            <h2>Products</h2>
            <p>As supplied by our vendor</p>
          </section>
          <section>
            <h2>
              <select onChange={this.props.loadProducts}>
                <option>AUD</option>
                <option>GBP</option>
                <option>USD</option>
              </select>
            </h2>
            <ReposList {...reposListProps} />
          </section>
          <section>
            <h2>Order</h2>
            <Order />
          </section>
        </div>
      </article>
    );
  }
}

HomePage.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.bool,
  ]),
  products: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.bool,
  ]),
  loadProducts: PropTypes.func,
};
