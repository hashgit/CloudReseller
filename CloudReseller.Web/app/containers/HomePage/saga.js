/**
 * Gets the repositories of the user from Github
 */

import { call, put, takeLatest, select } from 'redux-saga/effects';
import { productsLoaded, displayCurrency } from 'containers/HomePage/actions';

import request from 'utils/request';
import { LOAD_PRODUCTS } from 'containers/HomePage/constants';
import { makeSelectCurrency } from 'containers/HomePage/selectors';

/**
 */
export function* getProducts() {
  const currency = yield select(makeSelectCurrency());
  const requestURL = `${process.env.API_URL}api/products?currency=${currency}`;

  try {
    // Call our request helper (see 'utils/request')
    const products = yield call(request, requestURL);
    yield put(productsLoaded(products));
    yield put(displayCurrency(currency));
  } catch (err) {
    // yield put(repoLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* productsData() {
  yield takeLatest(LOAD_PRODUCTS, getProducts);
}
