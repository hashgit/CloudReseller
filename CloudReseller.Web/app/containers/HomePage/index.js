import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import {
  makeSelectProducts,
  makeSelectLoading,
  makeSelectError
} from 'containers/App/selectors';
import { loadProducts } from './actions';
import { makeSelectUsername } from './selectors';
import reducer from './reducer';
import saga from './saga';
import HomePage from './HomePage';

const mapDispatchToProps = (dispatch) => ({
  loadProducts: () => {
    dispatch(loadProducts());
  }
});

const mapStateToProps = createStructuredSelector({
  products: makeSelectProducts(),
  username: makeSelectUsername(),
  loading: makeSelectLoading(),
  error: makeSelectError()
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'home', reducer });
const withSaga = injectSaga({ key: 'home', saga });

export default compose(withReducer, withSaga, withConnect)(HomePage);
export { mapDispatchToProps };
