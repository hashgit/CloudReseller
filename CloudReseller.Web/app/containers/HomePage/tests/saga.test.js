/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_PRODUCTS } from 'containers/HomePage/constants';
import { productsLoaded, displayCurrency } from 'containers/HomePage/actions';

import productsData, { getProducts } from '../saga';

const currency = 'AUD';

/* eslint-disable redux-saga/yield-effects */
describe('getProducts Saga', () => {
  let getProductsGenerator;

  beforeEach(() => {
    getProductsGenerator = getProducts();

    const selectDescriptor = getProductsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getProductsGenerator.next(currency).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the productsLoaded action if it requests the data successfully', () => {
    const response = [{
      name: 'First product',
    }, {
      name: 'Second product',
    }];
    const putDescriptor = getProductsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(productsLoaded(response)));

    const currencyPutDescriptor = getProductsGenerator.next(currency).value;
    expect(currencyPutDescriptor).toEqual(put(displayCurrency(currency)));
  });
});

describe('productsData Saga', () => {
  const productsDataSaga = productsData();

  it('should start task to watch for LOAD_PRODUCTS action', () => {
    const takeLatestDescriptor = productsDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_PRODUCTS, getProducts));
  });
});
