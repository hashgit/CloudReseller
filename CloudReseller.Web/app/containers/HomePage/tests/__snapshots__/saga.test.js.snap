// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProducts Saga should dispatch the productsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getProducts Saga should dispatch the productsLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:61168/api/products?currency=AUD",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
